enhancements/insights/on-demand-data-gathering.md

```diff
@@ -0,0 +1,255 @@
+
+### Non-Goals
+
+TBD (or maybe N/A)
+
+## Proposal
+
+The enhancement proposal introduces a new Insights `CustomResourceDefinition` type in the OpenShift API that will serve as the definition and trigger to new data gathering on demand. 
+This new CRD will build on and extend the existing `GatherConfig` (see the [Insights config API enhancement](../insights/insights-config-api.md)). The idea is that with each new CR instance of this resource, new data gathering will happen. 
+
+The proposal for the new `datagather.insights.openshift.io` resource is:
+
+```go
+type DataGather struct {
+	metav1.TypeMeta   `json:",inline"`
+	metav1.ObjectMeta `json:"metadata,omitempty"`
+
+	// spec holds user settable values for configuration
+	// +kubebuilder:validation:Required
+	Spec DataGatherSpec `json:"spec"`
+	// status holds observed values from the cluster. They may not be overridden.
+	// +optional
+	Status DataGatherStatus `json:"status"`
+}
+
+type DataGatherSpec struct {
+	// dataPolicy allows user to enable additional global obfuscation of the IP addresses and base domain
+	// in the Insights archive data. Valid values are "None" and "ObfuscateNetworking".
+	// When set to None the data is not obfuscated.
+	// When set to ObfuscateNetworking the IP addresses and the cluster domain name are obfuscated.
+	// When omitted, this means no opinion and the platform is left to choose a reasonable default, which is subject to change over time.
+	// The current default is None.
+	// +optional
+	DataPolicy DataPolicy `json:"dataPolicy,omitempty"`
+	// disabledGatherers is a list of gatherers to be excluded from the gathering. All the gatherers can be disabled by providing "all" value.
+	// If all the gatherers are disabled, the Insights operator does not gather any data.
+	// The particular gatherers IDs can be found at https://github.com/openshift/insights-operator/blob/master/docs/gathered-data.md.
+	// Run the following command to get the names of last active gatherers:
+	// "oc get insightsoperators.operator.openshift.io cluster -o json | jq '.status.gatherStatus.gatherers[].name'"
+	// An example of disabling gatherers looks like this: `disabledGatherers: ["clusterconfig/machine_configs", "workloads/workload_info"]`
+	// +optional
+	DisabledGatherers []string `json:"disabledGatherers"`
+}
+
+const (
+	// No data obfuscation
+	NoPolicy DataPolicy = "None"
+	// IP addresses and cluster domain name are obfuscated
+	ObfuscateNetworking DataPolicy = "ObfuscateNetworking"
+	// Data gathering is running
+	Running State = "Running"
+	// Data gathering is completed
+	Completed State = "Completed"
+	// Data gathering failed
+	Failed State = "Failed"
+)
+
+// dataPolicy declares valid data policy types
+// +kubebuilder:validation:Enum="";None;ObfuscateNetworking
+type DataPolicy string
+
+// state declares valid gathering state types
+// +kubebuilder:validation:Enum=Running;Completed;Failed
+type State string
```

Comment:
Hmm, would that basically mean that on-demand gathering requests are queued?

Reply:
Yes, I guess we have to make a decision on whether they are queued or not, seems that's maybe not what you want?

But if one gets interrupted because a newer one is created, does this mean it goes into the `Failed` state? I wonder then if it's better to have `Superseded`/`Cancelled` as a state to indicate this is not a regular failure. Depends on which way we go with the queueing vs interruption behaviour

Reply:
Yes I agree. This is a crucial decision that we must make first in my opinion. I think I have two points with regard the queue approach:
- my main concern is possibly easy way to overwhelm our console.redhat.com ingress (someone creates hundreds of `DataGather` CRs and the ingress will get a new archive every few minutes. Compared to each 2 hours right now). Maybe we could limit the queue somehow...not sure.
- I like the fact that it avoids the parallelism while preserving the latest result of the data gathering
  -  data gathering A is running
  - data gathering B is queued, but it waits until A finishes and writes the result (including the Insights analysis from console.redhat.com) to the new CR

Reply:
Agree. Let me circle back to the discussion about the `state` attribute. I wonder whether it would make sense to use `conditions` instead, because if, for example,  an upload fails, then we need to know reason, message etc. to propagate to the clusteroperator conditions.

Reply:
This was already resolved. Conditions are part of the status.